jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java

```diff
@@ -307,66 +345,220 @@ private void closeOperationHandle(TOperationHandle stmtHandle) throws SQLExcepti
+                    stmtHandle, orientation, fetchSize);
+                TFetchResultsResp fetchResp = fetchClient.FetchResults(fetchReq);
+                Utils.verifySuccessWithInfo(fetchResp.getStatus());
+                TRowSet results = fetchResp.getResults();
+                if (results.getStartRowOffset() > 0) {
+                  hasStartRow = true;
+                }
+                result.fetchedRows = RowSetFactory.create(results, protocol);
+                result.numRows = result.fetchedRows.numRows();
+                boolean hasMoreRows = result.numRows > 0 || fetchResp.isHasMoreRows();
+                if (!hasMoreRows) {
+                  gotLastBatch = true;
+                }
+                result.hasMoreRows = hasMoreRows;
+                result.fetchedRows = RowSetFactory.create(results, protocol);
+                if (useConvertedResultSet) {
+                  result.fetchedRows =
+                      new ConvertedResultSet(result.fetchedRows, getSchema());
+                }
+                result.startRow = results.getStartRowOffset() + 1;
+                if (hasStartRow && result.startRow < nextStartRow.get()) {
+                  throw new SQLException("Unexpected row offset");
+                }
+              } catch (Exception e) {
+                result.ex = e;
+              }
+
+              try {
+                // Wait for earlier row sets to be added to the queue
+                synchronized(nextStartRow) {
+                  if (!poolDone) {
+                    if (result.ex == null) {
+                      if (hasStartRow) {
+                        while (nextStartRow.get() != result.startRow) {
+                          nextStartRow.wait();
+                        }
+                        nextStartRow.set(result.startRow + result.numRows);
+                      }
+                      poolDone = !result.hasMoreRows;
+                    } else {
+                      poolDone = true;
+                    }
+                    resultQueue.put(result);
+                    if (hasStartRow) {
+                      nextStartRow.notifyAll();
+                    }
+                  }
+                }
+              } catch (InterruptedException e) {
+                interruptException = e;
+                break;
+              }
+            }
+            // TODO: shutdown called here can make execute return
+            // RejectedExecutionException
+            // pool.shutdownNow();
+          });
+        }
+      }
+      try {
+        if (interruptException != null) {
+          throw interruptException;
+        }
+        FetchResult result = resultQueue.take();
+        fetchDone = !result.hasMoreRows;
+        if (result.ex != null) {
+          ShutdownPool();
+          throw new SQLException(result.ex);
+        }
+        fetchedRows = result.fetchedRows;
+      } catch (InterruptedException e) {
```

Comment:
should we `shutdown` the pool in `finally` block?

Reply:
Not here. This is just early shutdown for error case. Unconditional shutdown is from close()